Consolidate HexPrinter and BytePrinter

zhenlineo opened the pull request:
Removed the redundant BytePrinter.
Extended the existing HexPrinter class to be able to control the length of line number (none, 1, 01, etc.) and byte group separator ("  ", "\t", etc.).
Added a few static methods in HexPrinter to make the use of it easier.

thobe commented:
How about incorporating the `dumpHex`-method of `DumpStore` as well, since it also does hex printing...

zhenlineo commented:
@thobe They use quite different formatting:
`DumpStore#dumpHex` uses hex line number which starts at an arbitrary number, arranged in groups of four bytes.

```
@ 0x0000000a:  00 01 02 03  04 05 06 07  08 09 0a 0b  0c 0d 0e 0f
@ 0x0000001a:  10 11 12 13  14 15 16 17  18 19 1a 1b  1c 1d 1e 1f
```

`BytePrinter` uses decimal line number which starts from fixed 0, arranged in groups of 8 bytes.

```
00000000 00 01 02 03 04 05 06 07  08 09 0a 0b 0c 0d 0e 0f
00000001 10 11 12 13 14 15 16 17  18 19 1a 1b 1c 1d 1e 1f
```

It might be too complex for a class to handle both of them.

thobe commented:
@zhenlineo I don't think that difference is important enough, let `DumpStore` use the same format as `HexPrinter`/`BytePrinter`.

zhenlineo commented:
Okay, that should be easy to handle. Only need to introduce a `lineNumberOffset` then.

cleishm commented:
There's also https://github.com/neo4j/neo4j/blob/2.3/community/kernel/src/main/java/org/neo4j/kernel/impl/util/Codecs.java

Both `HexPrinter` and `BytePrinter` seem poor names, given they don't conform to `Printer`...

zhenlineo commented:
@thobe No, `DumpStore` is doing something really smart. The line number has to be a hex number to cover all values in `long`. I would keep it as original as I like its original formatting.

@cleishm `Codecs` is doing something totally different. Even though they all deal with hex numbers, the  output format is totally different. I do not think it is necessary that we have a super powerful class to handle all possible formatting of hex printing.

Commit:
Extended HexPrinter to handle multiple header style, grouping style.

Now we standardize the format to print multiple hex strings:
 - The line numbers are always coded in hex.
 - Each byte is presented in two hex digits and bytes are separated with one space.

However one could freely choose how many digits to use for line numbers, what suffix or prefix around line number digits, how many bytes in each line and each group, and what group separator to use.

jakewins commented:
@zhenlineo how about changing HexPrinter to use the print format used by DumpStore then? DumpStore format seems like the better of the two options. This PR is a good opportunity to clear a lot of this code up - deleting stuff ftw!

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/DumpStore.java:
```diff
@@ -291,64 +292,45 @@ private boolean dumpRecord( STORE store, int size, StoreChannel fileChannel, Byt
-        int pos = buffer.position();
-        try
+        HexPrinter printer = new HexPrinter( out ).withBytesGroupingFormat( 16, 4, "  " ).withLineNumberDigits( 8 )
+                .withLineNumberOffset( id * size );
```
Do we really need to instantiate and configure a new `HexPrinter` for each record?

zhenlineo replied:
Good point. Pulled out as a field.

thobe commented:
LGTM
